### PR TITLE
add transform data functionality

### PR DESCRIPTION
**Pull request by sanjmohan**
This adds functionality to transform data with UMAP with respect to previously fit data.

See the [README additions](https://github.com/dillondaudert/UMAP.jl/compare/master...sanjmohan:feat/transform-data?expand=1#diff-04c6e90faac2675aa89e2176d2eec7d8) for usage overview.

First, a skeleton is build for the manifold containing both the old fit data and the new data to transform. This yields a graph whose first R rows/columns correspond to the old fit data (for which an embedding has already been obtained), and the remaining rows/columns correspond to the new data to be transformed. This uses the preexisting `fuzzy_simplicial_set` function as is.

Next, an embedding for the data to transform is initialized as a weighted mean of points from the reference embedding. This is the new method overload for the `initialize_embedding` function. For comparison, [here is the python implementation](https://github.com/lmcinnes/umap/blob/96f1291fb5ba70df18f67fffb588bd7cbcabee5e/umap/umap_.py#L1126) (the weights are obtained from the graph entries, which you can see [here](https://github.com/lmcinnes/umap/blob/96f1291fb5ba70df18f67fffb588bd7cbcabee5e/umap/umap_.py#L2029) ).

Finally, the `optimize_embedding` function operates as before, but only computes attractions / repulsions between the new embedding and the reference embedding. For the python version, see [here](https://github.com/lmcinnes/umap/blob/96f1291fb5ba70df18f67fffb588bd7cbcabee5e/umap/layouts.py#L243) They use `head` to refer to the new embedding, which I designated with `query_inds`, and `tail` to refer to the reference embedding, which I designated with `ref_inds`.

When using this via a high level `umap` call, the reference (fit) data is assumed to be the first R samples of the input data. However, the `optimize_embedding` function can refer to arbitrary indices of the input graph / input embedding as the reference data, and arbitrary indices as the query data (data to be optimized/transformed). These indices don't necessarily have to cover all indices of the graph/embedding (the remaining entries will be ignored entirely). They can also overlap (which means that data may be optimized by computing attractions/repulsions to other data that is also intended to be optimized).

I also added a few basic tests to cover these new features.

Let me know what you think about this addition, or if there are other parts I can clarify (either in this PR or documentation)!

### CODE REVIEW

**Comment by sanjmohan** [comment]
I also considered a different approach to the high level API, which was letting the `UMAP_` struct hold all the parameters from its fit call, and reuse these exact parameters to transform new data. This would require the struct holding the values or a reference to the original data along with the original embedding -- since a new graph encompassing old and new data must be made -- and other hyperparameters. This would be more similar to the Python OOP approach that the author of the paper has written. This would require more refactoring of the code and might be more restrictive in terms of exposing functionality, but also might simplify the API. I went with the approach here since it has a smaller footprint on the existing code. But I am willing to reconsider :)

**Comment by dillondaudert** [comment]
Thanks for the PR, this looks great! I will give it a review in the next couple days.

**Comment by sanjmohan** [comment]
fixed docstring typos

**Comment by dillondaudert** [comment]
I took an initial look through and this seems mostly on the right track. I need to spend more time with this, so I'll have a more thorough review forthcoming.

There is one potential issue with when optimizing a new embedding with respect to the reference. Since the knn graph is constructed from the union of the reference and query data, it leaves open the possibility that some query point has no neighbors (either incoming or outgoing) in the reference data. During optimization, this means that its embedding will never be updated (the condition at [line 155](https://github.com/dillondaudert/UMAP.jl/pull/26/files#diff-f98aa7325db52ead855f818e3e2ae2f3R155) in embeddings.jl will always be false.)

I wonder if you've thought about this and how the Python implementation handles it, if it does.

**Comment by sanjmohan** [comment]
Ah you're right, I didn't consider this. Looks like the Python implementation may actually build the graph by only searching for nearest neighbors in the reference data. I'll look into this more and update the PR within a few days.

**Comment by sanjmohan** [comment]
As I am looking through the Python implementation and the codebase here and in NearestNeighborDescent.jl, I'm having a bit of a change of heart about API. I'm now thinking that what may make the most sense is the use case
```julia
# easy option: quick data --> embedding (the current api)
embedding = umap(X, n_components, kwargs...)

# and more exposed option, allowing transforms of new data into preexisting embedded space:
model = UMAP_(X, n_components, kwargs...)
embedding = model.embedding
test_embedding = umap(Xtest, model, kwargs...)
```

The upsides are:
- we can then reuse the ApproximateKNN graph from the original embedding to use the `search` from NearestNeighborDescent.jl to get neighbors of query data in the original dataset. The Python implementation does some partial ApproximateKNN graph construction sort of computation [here](https://github.com/lmcinnes/umap/blob/fdd1096ce6086e72369d8280bb5a016cc5ee5e30/umap/nndescent.py#L265) and [here](https://github.com/lmcinnes/umap/blob/fdd1096ce6086e72369d8280bb5a016cc5ee5e30/umap/nndescent.py#L276) using the original data but it seems to me that one of the only differences here is that the number of neighbors isn't restricted to a precomputed neighbor graph, but they enforce using the same number of neighbors anyways
- we can then restrict the options / kwargs in the transform api and fix the others to those used in the first manifold construction embedding (tracked by the `UMAP_` struct)

The only drawback of this from the user POV is that you can't transform new data w.r.t a dataset and embedding without its neighbor graph, eg. if they were obtained externally, but I feel like this won't be a use case for most people, and even if so, recomputing the neighbor graph of the data won't be prohibitive in most cases.

The main changes then are just exposing the neighbor graph made by the `knn_search` functions to the UMAP_ struct, and then supporting use of the UMAP_ struct moreso than before.

I will update the PR with this version soon but let me know if you have any objections to this approach! :)

**Comment by dillondaudert** [comment]
I agree that we'll need to store the state from the initial fit procedure in order to use it for the transform. 

This will require some changes to the `UMAP_` struct, in particular I think we'll need to add `indices` and `distances` from the `knn_search` function as struct fields (so this call will need to be pulled out of `fuzzy_simplicial_set`). I don't think we should store the `KNNGraph` directly, since sometimes we calculate neighbors using an exact method, and we can reconstruct the knn graph using these fields later.

We'll also need add the initial data used as a field so we can calculate the distances during the transform.

Then there will need to be a new method of `knn_search` that does something like reconstructs the KNNGraph `HeapKNNGraph(data, metric, indices, distances)` and calls the `search` function to get the inds/dists for the new data to be transformed `query_inds, query_dists = search(knngraph, queries, n_neighbors)`.

So I'm alright with a bit of refactoring as needed, especially in the `UMAP_` struct since that was never user-facing anyway. I have an idea for how to support the case where the user passes precomputed distances, but you can ignore that for now. Let me know if you have any questions as you're working through this.

**Comment by sanjmohan** [comment]
Hi! I've gone through and implemented the things you suggested, and changed the transform function (now called `umap_transform`) to work with a new `UMAP_` struct. The new (way simpler) usage info is in the new README. Let me know what you think!

One thing I discovered, that may need to factor into this PR or maybe a next PR:

It seems like the python implementation has been optimized a lot recently. I ran the notebook in the repo, and according to the README, python umap was a bit slower than this package. Now, this package has similar timing to the README (~150sec) while the python umap runs in only ~50 seconds on my laptop. It seems like one of the major areas of optimization was their improvement of the nearest neighbor search, which can be looked at later.

The side effect though is that I was comparing the python functionality to transforming new data to this PR, and got the following results on mnist test data, after fitting to training data:

Julia:
image

Python:
image

So using the hyperparameters, etc. in this PR, by the eye check it seems like the Python transform is more accurate. It also runs in about 10 seconds while the julia function takes ~25 seconds -- so I was not sure how to compare results while keeping a similar level of nearest neighbor approximation complexity, since the python has a faster baseline as of now. Increasing the number of neighbors used in the ApproximateKNNGraph made the julia transform plot more closely resemble the python plot.

**Comment by dillondaudert** [comment]
Hi - thanks for the update, I'll take another look over the weekend!

You're right about pynndescent being faster now than when I created the readme, my impression was that it was mostly due to parallelization. I looked into adding the same to NNDescent.jl but I remember I wasn't satisfied with the approach they took.

File: src/embeddings.jl [hunk 1]
```diff
@@ -19,6 +19,24 @@ function initialize_embedding(graph::AbstractMatrix{T}, n_components, ::Val{:ran
     return [20 .* rand(T, n_components) .- 10 for _ in 1:size(graph, 1)]
 end
 
+"""
+    initialize_embedding(graph::AbstractMatrix{<:Real}, ref_embedding::AbstractMatrix{T<:AbstractFloat}) -> embedding
+
+Initialize an embedding of points corresponding to the columns of the `graph`, by taking weighted means of
+the columns of `ref_embedding`, where weights are values from the rows of the `graph`.
+
+The resulting embedding will have shape `(size(ref_embedding, 1), length(query_inds))`, where `size(ref_embedding, 1)`
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/embeddings.jl:28]: ah mb, forgot to update this documentation

File: src/embeddings.jl [hunk 2]
```diff
@@ -53,22 +78,71 @@ Optimize an embedding by minimizing the fuzzy set cross entropy between the high
...
-- `n_epochs`: the number of training epochs for optimization
+- `n_epochs`::Integer: the number of training epochs for optimization
 - `initial_alpha`: the initial learning rate
 - `gamma`: the repulsive strength of negative samples
 - `neg_sample_rate::Integer`: the number of negative samples per positive sample
 """
 function optimize_embedding(graph,
-                            embedding,
-                            n_epochs,
+                            embedding::AbstractVector{<:AbstractVector{<:AbstractFloat}},
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/embeddings.jl:87]: So one issue was having the `optimize_embedding` methods with both one embedding argument and with two (for query and ref), the call would not be properly dispatched because of the default arguments to `_a` and `_b`, and I was resolving this with type constraints. But I can remove the default arguments or remove the overloaded methods, if you have a preference towards either of those?

File: src/umap_.jl [hunk 3]
```diff
@@ -1,24 +1,43 @@
...
+        size(knns) == size(dists) || error("UMAP_ constructor expected knns and dists to have equal size")
+        new(graph, embedding, data, knns, dists)
+    end
+    # For backwards compatibility:
     function UMAP_{S, M, N}(graph, embedding) where {S<:Real,
                                                      M<:AbstractMatrix{S},
                                                      N<:AbstractMatrix{S}}
-        issymmetric(graph) || isapprox(graph, graph') || error("UMAP_ constructor expected graph to be a symmetric matrix")
-        new(graph, embedding)
+        UMAP_{S, M, N}(graph, embedding, Matrix(undef, 0, 0), Matrix{Int}(undef, 0, 0), Matrix{S}(undef, 0, 0))
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/umap_.jl:22]: One thing I was trying to do was leave old constructors since `UMAP_` had been exported before, so these changes wouldn't break old code. But since the struct didn't have much of a use before this may not be a big concern.

The other thing I was considering was I have had use cases where I have data and its corresponding embedding, although the embedding was not computed from the same package. I then wanted to use some `transform` sort of functionality, and recomputing the graph of the old data would be redundant. Constructing a partially defined struct would let a user wrap data and its embedding manually, and use these for transforming new data.
(As is, this wouldn't save a whole lot of computation since the current knn_search would recompute the ApproximateKNNGraph for large matrices anyways, but this would be an internal detail that could change later). But let me know your thoughts on this!

File: src/umap_.jl [hunk 4]
```diff
@@ -1,24 +1,43 @@
 # an implementation of Uniform Manifold Approximation and Projection
 # for Dimension Reduction, L. McInnes, J. Healy, J. Melville, 2018.
 
-# NOTE: unused for now
 struct UMAP_{S <: Real, M <: AbstractMatrix{S}, N <: AbstractMatrix{S}}
     graph::M
     embedding::N
+    data::AbstractMatrix
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/umap_.jl:7]: Is the intended use-case for this package to just operate on real numbers? I see that the original umap function requires real matrices, but it seems like other packages handle more general data, and one of the advantages of the umap algorithm is its ability to use arbitrary metrics to embed manifolds in arbitrary spaces into low-dim Euclidean space. Do you plan to do something like this in the future?

File: src/umap_.jl [hunk 5]
```diff
@@ -75,42 +109,115 @@ function UMAP_(X::AbstractMatrix{S},
...
+    0 ≤ set_operation_ratio ≤ 1                      || throw(ArgumentError("set_operation_ratio must lie in [0, 1]"))
+    local_connectivity > 0                           || throw(ArgumentError("local_connectivity must be greater than 0"))
+    !isempty(model.data)                             || throw(ArgumentError("model.data must not be empty"))
+    size(model.data, 2) == size(model.embedding, 2)  || throw(ArgumentError("model.data must have same number of columns as model.embedding"))
+    size(model.data, 1) == size(Q, 1)                || throw(ArgumentError("size(model.data, 1) must equal size(Q, 1)"))
+
+
+    n_epochs = max(0, n_epochs)
+    # main algorithm
+    knns, dists = knn_search(model.data, Q, n_neighbors, metric, model.knns, model.dists)
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/umap_.jl:176]: The verbosity would just be moved from one function to another, and this would additionally introduce some (logical) circular dependencies between "umap_.jl" (which defines the model) and "utils.jl" (which defines the `knn_search`, and has been so far loaded before "umap_.jl". The actual circular dependency could easily be hidden by removing any type constraints from the `knn_search` parameter and it would work perfectly fine but this just seems a bit more hacky to me... But overall this just seems like personal preference and I can still change this if you'd like.

File: src/embeddings.jl [hunk 6]
```diff
@@ -19,6 +19,24 @@ function initialize_embedding(graph::AbstractMatrix{T}, n_components, ::Val{:ran
...
+Initialize an embedding of points corresponding to the columns of the `graph`, by taking weighted means of
+the columns of `ref_embedding`, where weights are values from the rows of the `graph`.
+
+The resulting embedding will have shape `(size(ref_embedding, 1), length(query_inds))`, where `size(ref_embedding, 1)`
+is the number of components (dimensions) of the `reference embedding`, and `length(query_inds)` is the number of 
+samples in the resulting embedding. Its elements will have type T.
+"""
+function initialize_embedding(graph::AbstractMatrix{<:Real}, ref_embedding::AbstractMatrix{T}) where {T<:AbstractFloat}
+    embed = [zeros(T, size(ref_embedding, 1)) for _ in 1:size(graph, 2)]
+    for (i, col) in enumerate(eachcol(graph))
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/embeddings.jl:34]: good catch

File: src/embeddings.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/embeddings.jl:87]: If you have no preference, I will just remove the original `optimize_embedding` method so all calls will have to pass two embeddings (or the same one twice), I'm starting to think the overloading is more unnecessary/messy than good.

File: src/utils.jl [hunk 7]
```diff
@@ -69,8 +89,31 @@ function knn_search(X::AbstractMatrix{S},
...
+                    k::Integer,
+                    metric::SemiMetric,
+                    knns::AbstractMatrix{<:Integer},
+                    dists::AbstractMatrix{<:Real})
+
+    if size(X, 2) < 4096
+        return _knn_from_dists(pairwise(metric, X, Q, dims=2), k, ignore_diagonal=false)
+    end
+
+    if isempty(knns) || isempty(dists) || size(knns) != size(dists)
```
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/utils.jl:103]: I addressed this in reply to your other knn_search comment :+1:

File: src/umap_.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (dillondaudert) [reply src/umap_.jl:176]: It's not a huge issue, so you can leave the arguments as they are for now.

File: src/umap_.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (dillondaudert) [reply src/umap_.jl:7]: Yes, it's definitely something that I've been wanting to do. I've put some thought into how to redesign the internals in order to handle transforming from one arbitrary topological space to another, but I don't have a timeline on that.

File: src/umap_.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (dillondaudert) [reply src/umap_.jl:22]: Maybe you could give me an example of the workflow when you'd need a partial struct? I'm not sure I understand the use case entirely.

I wouldn't worry too much about "breaking" API changes - since this package is pre-1.0, any minor version bump technically permits breaking changes. There's still a ways to go before a 1.0 release is ready and the API could be made stable.

[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/umap_.jl:22]: Ok, I won't worry about the API changes :+1:

The use case was that a third party published data and its embedding, and I had some related data of interest that I wanted to embed along with the published data/embedding. Reconsidering this (and since we are changing the struct API anyways), I think a separate `transform` method would be much better than the partial struct.

File: src/utils.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/utils.jl:103]: I am removing this if/else

File: src/umap_.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (sanjmohan) [reply src/umap_.jl:7]: thanks for the tips here, still learning about Julia's type system

**Comment by sanjmohan** [comment]
I believe I addressed all of your comments and made the changes. The major ones were:
- adding type parameterization to the `UMAP_` struct
- getting rid of the old `optimize_embedding` method, so the only one is the one with both reference and query embedding arguments
- cleaning up the `initialize_embedding` function with your suggestions
- cleaning up the `knn_search` methods

The `transform` function should be type stable now from begiinning to end, returning an embedding of the same type as the embedding in the model argument.

Here are the relative benchmarks:

MASTER
image

NEW BRANCH
image

(the data is the typical mnist data, (28*28, 60k) for `mnist_x` and (28*28, 10k) for `mnist_xtest`)

**Comment by dillondaudert** [comment]
Thanks for the quick work! I'll aim to get a final review by the weekend, and hopefully we can get this merged soon after that.

**Comment by sanjmohan** [comment]
cool, if you'd like to open some issues I can help out!
